packages/stargate/src/feemarket.ts

```diff
@@ -0,0 +1,207 @@
+import { Decimal } from "@cosmjs/math";
+import { BinaryReader, BinaryWriter } from "cosmjs-types/binary";
+
+import { GasPrice } from "./fee";
+import { decodeCosmosSdkDecFromProto } from "./queryclient/utils";
+
+/**
+ * Configuration for dynamic gas price queries from feemarket modules.
+ */
+export interface DynamicGasPriceConfig {
+  /** Denomination to query gas price for (e.g., "uatom") */
+  readonly denom: string;
+  /** Multiplier to apply to the queried gas price (e.g., 1.1 for 10% above base). Defaults to 1.3 */
+  readonly multiplier?: number;
+  /** Minimum gas price to use as fallback if query fails and as a floor for the dynamic price */
+  readonly minGasPrice: GasPrice;
+  /** Maximum gas price to cap at. If not provided, no maximum is enforced */
+  readonly maxGasPrice?: GasPrice;
+}
+
+/**
+ * Multiplies a Decimal by a number multiplier.
+ * This is needed because Decimal.multiply() only accepts integers.
+ */
+export function multiplyDecimalByNumber(
+  value: Decimal,
+  multiplier: number,
+  fractionalDigits: number,
+): Decimal {
+  const normalizedValue =
+    value.fractionalDigits === fractionalDigits
+      ? value
+      : Decimal.fromUserInput(value.toString(), fractionalDigits);
```

Comment:
This part is something the `class Decimal` should have. I.e. convert decimal "2.0" to "2.000" by changing the number of decimals from 1 to 3